### PR TITLE
layers schema

### PR DESCRIPTION
**Pull request by thewizarodofoz**
Closes: #8015

### CODE REVIEW

File: lib/configSchema.js [hunk 1]
```diff
@@ -96,22 +109,42 @@ const schema = {
...
-        excludeDevDependencies: { type: 'boolean' },
+      additionalProperties: {
+        type: 'object',
+        properties: {
+          allowedAccounts: {
+            type: 'array',
+            items: {
+              oneOf: [
+                { type: 'integer', minimum: 100000000000, maximum: 999999999999 },
+                { type: 'string', pattern: '^\\d{12}$' },
```
Comment (medikoo) [inline lib/configSchema.js:124]: I think pattern should be as here: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-lambda-layerversionpermission.html#cfn-lambda-layerversionpermission-principal

File: lib/configSchema.js [hunk 2]
```diff
@@ -96,22 +109,42 @@ const schema = {
...
+                { const: '*' },
```
Comment (medikoo) [inline lib/configSchema.js:125]: I would remove it and rely on pattern as in AWS docs

**Review by medikoo — CHANGES_REQUESTED** [review]
Thank you @thewizarodofoz ! It'll be great to have that in, and it looks very promising, please see my remarks

File: lib/configSchema.js [hunk 3]
```diff
@@ -96,22 +109,42 @@ const schema = {
       additionalProperties: false,
     },
 
-    package: {
+    package: packageSchema,
+
+    layers: {
```
[the remark this answers is not in the document]
Comment (thewizarodofoz) [reply lib/configSchema.js:114]: So should I remove `layers` from `lib/configSchema.js` altogether? or keep the simple `{type: 'object'}` schema there?

[the remark this answers is not in the document]
Comment (medikoo) [reply lib/configSchema.js:114]: I believe it should be moved completely, exactly in a same way as `resources` where moved with this commit: https://github.com/serverless/serverless/commit/6d7e96722721c8a5c614bea2802af7142011d35f

File: lib/plugins/aws/provider/awsProvider.js [hunk 4]
```diff
@@ -504,6 +504,53 @@ class AwsProvider {
...
+            type: 'object',
+            properties: {
+              allowedAccounts: {
+                type: 'array',
+                items: {
+                  oneOf: [
+                    { type: 'integer', minimum: 100000000000, maximum: 999999999999 },
+                    {
+                      type: 'string',
+                      pattern: '\\d{12}|\\*|arn:(aws[a-zA-Z-]*):iam::\\d{12}:root',
```
Comment (medikoo) [inline lib/plugins/aws/provider/awsProvider.js:519]: I think we need to enclose it with `^` and `$`

**Review by medikoo — CHANGES_REQUESTED** [review]
@thewizarodofoz thanks for update! It looks promising, still if I see correctly, in current form this schema is totally ignored.

To have it working,  we need to introduce exactly same changes as we did here https://github.com/serverless/serverless/commit/6d7e96722721c8a5c614bea2802af7142011d35f for `resources`
(note changes in  `lib/classes/ConfigSchemaHandler/index.js` and I think it'll be nice to add documentation line at  `docs/providers/aws/guide/plugins.md`)

**Comment by thewizarodofoz** [comment]
> I think it'll be nice to add documentation line at `docs/providers/aws/guide/plugins.md`

@medikoo Not sure that I follow. If I understand correctly, `resources` is mentioned there in the context of "Definition for eventual top level "resources" section", but in this PR we said that `layers` is not a top level section (it's aws specific) so it will not be relevant to a new custom provider.

**Comment by thewizarodofoz** [comment]
@medikoo I followed the same path as shown in 6d7e96722721c8a5c614bea2802af7142011d35f, hope it's fine.
If you can, I would like to know why those are needed, couldn't figure it out from the code.

**Review by medikoo — COMMENTED** [review]
> @medikoo Not sure that I follow. If I understand correctly, resources is mentioned there in the context of "Definition for eventual top level "resources" section", but in this PR we said that layers is not a top level section (it's aws specific) so it will not be relevant to a new custom provider.

`layers` actually remain top-level, we didn't change that. We just moved it's definition to AWS context.

In this PR we move it's definition, and a capability to define `layers` through `defineProvider` schema extension utility, and to have that fully complete, it'll be great to also expose in docs that `layers` can be defined via `defineProvider`. See full docs for `defineProvider`: https://github.com/serverless/serverless/blob/6d7e96722721c8a5c614bea2802af7142011d35f/docs/providers/aws/guide/plugins.md#defineprovider-helper

File: lib/plugins/aws/provider/awsProvider.js [hunk 5]
```diff
@@ -504,6 +504,53 @@ class AwsProvider {
...
+        layers: {
+          type: 'object',
+          additionalProperties: {
+            type: 'object',
+            properties: {
+              allowedAccounts: {
+                type: 'array',
+                items: {
+                  oneOf: [
+                    { type: 'integer', minimum: 100000000000, maximum: 999999999999 },
```
Comment (medikoo) [inline lib/plugins/aws/provider/awsProvider.js:516]: Today I've turned on automatic coercion with https://github.com/serverless/serverless/pull/8319, and having that I believe we  no longer need that `number` definition.

We can also remove the logic which tries to handle eventual numeric input

**Review by medikoo — COMMENTED** [review]
That looks great @thewizarodofoz !

Sorry for still not yet taking it in, but today one improvement landed which I believe will vastly improve service configuration handling, and which allows us to simply schema as proposed here. Please see my comment

**Review by medikoo — APPROVED** [review]
Thank you @thewizarodofoz ! That looks great

**Pull request merged**